libraries/AP_Mount/AP_Mount_Gremsy.cpp

```diff
@@ -20,76 +20,92 @@ void AP_Mount_Gremsy::update()
-            send_gimbal_device_set_attitude(ToRad(angle_bf_target.x), ToRad(angle_bf_target.y), ToRad(angle_bf_target.z), false);
-            }
+            mnt_target.target_type = MountTargetType::ANGLE;
+            mnt_target.angle_rad.roll  = ToRad(angle_bf_target.x);
+            mnt_target.angle_rad.pitch = ToRad(angle_bf_target.y);
+            mnt_target.angle_rad.yaw   = ToRad(angle_bf_target.z);
+            mnt_target.angle_rad.yaw_is_ef = false;
+            isTargetValid = true;
             break;
+        }
 
-        // use angle or rate targets provided by a mavlink message or mission command
-        case MAV_MOUNT_MODE_MAVLINK_TARGETING:
-            switch (mavt_target.target_type) {
-            case MountTargetType::ANGLE:
-                send_gimbal_device_set_attitude(mavt_target.angle_rad.roll, mavt_target.angle_rad.pitch, mavt_target.angle_rad.yaw, mavt_target.angle_rad.yaw_is_ef);
-                break;
-            case MountTargetType::RATE:
-                send_gimbal_device_set_rate(mavt_target.rate_rads.roll, mavt_target.rate_rads.pitch, mavt_target.rate_rads.yaw, mavt_target.rate_rads.yaw_is_ef);
-                break;
-            }
+        case MAV_MOUNT_MODE_MAVLINK_TARGETING: {
+            // mavlink targets are stored while handling the incoming message and are considered valid
+            isTargetValid = true;
             break;
+        }
 
         // RC radio manual angle control, but with stabilization from the AHRS
         case MAV_MOUNT_MODE_RC_TARGETING: {
             // update targets using pilot's rc inputs
-            MountTarget rc_target {};
-            if (get_rc_rate_target(rc_target)) {
-                send_gimbal_device_set_rate(rc_target.roll, rc_target.pitch, rc_target.yaw, rc_target.yaw_is_ef);
-            } else if (get_rc_angle_target(rc_target)) {
-                send_gimbal_device_set_attitude(rc_target.roll, rc_target.pitch, rc_target.yaw, rc_target.yaw_is_ef);
+            mnt_target.target_type = MountTargetType::RATE;
+            isTargetValid = get_rc_rate_target(mnt_target.rate_rads);
+            if (!isTargetValid) {
+                mnt_target.target_type = MountTargetType::ANGLE;
+                isTargetValid = get_rc_angle_target(mnt_target.angle_rad);
             }
+            mnt_target.angle_rad.yaw_is_ef = false;
             break;
         }
 
         // point mount to a GPS point given by the mission planner
         case MAV_MOUNT_MODE_GPS_POINT: {
-            MountTarget angle_target_rad {};
-            if (get_angle_target_to_roi(angle_target_rad)) {
-                send_gimbal_device_set_attitude(angle_target_rad.roll, angle_target_rad.pitch, angle_target_rad.yaw, angle_target_rad.yaw_is_ef);
-            }
+            mnt_target.target_type = MountTargetType::ANGLE;
+            mnt_target.angle_rad.yaw_is_ef = false;
+            isTargetValid = get_angle_target_to_roi(mnt_target.angle_rad);
```

Comment:
I think it would be better to only change the mnt_target.target_type and mnt_target.angle_rad.yaw_is_ef if get_angle_target_to_roi() returns true.  So something like this:

```
if (get_angle_target_to_roi(mnt_target.angle_rad)) {
    mnt_target.target_type = MountTargetType::ANGLE;
    mnt_target.angle_rad.yaw_is_ef = false;
    isTargetValid = true;
}

```